### DISCUSSION TITLE
Late provider and null safety

### DISCUSSION THREAD

**Discussion by scopendo** [rrousselGit/riverpod#484] [Q&A] [answered]
Hi – I have a number of cases where a provider is being used like a `late` variable, meaning that its value is going to be set before it is first read (or watched in Riverpod speak).  The downside of this is that I have to make the value type nullable, which leads to a noisy use of `!` to avoid null checks.

**Case 1**

This is a provider whose value I will set before it is read/watched.

```dart
final confirmationProvider = StateProvider.autoDispose<AsyncValue<Confirmation>?>(
  (ref) => null,
);

// Before access of this provider I know that the code will set the provider's state
context.read(confirmationProvider).state = confirmation;
```

In this case, I could choose to initialise the value with an `AsyncLoading` instance but I haven't because that's not actually true.  Perhaps this is the pragmatic choice but it doesn't address case 2.

**Case 2**

This is a provider that gives access to the `SharedPreferences` instance without needing to worry about async code.  It would be preferable to have this provider also be non-nullable.

```dart
final sharedPreferencesProvider = Provider<SharedPreferences?>((_) => null);

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final sharedPreferences = await SharedPreferences.getInstance();
  runApp(
    ProviderScope(
      overrides: [
        sharedPreferencesProvider.overrideWithValue(sharedPreferences),
      ],
      child: MyApp(),
    ),
  );
}
```

I wonder if there's a late-like approach whereby a provider can be declared without a create function, requiring its null safe value to be set before it is read/watched?

**Comment by rrousselGit** [comment, marked as answer]
For case 2, have the default implementation throw instead:

```dart
final sharedPreferencesProvider = Provider<SharedPreferences>((_) => throw UnimplementedError());

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  final sharedPreferences = await SharedPreferences.getInstance();
  runApp(
    ProviderScope(
      overrides: [
        sharedPreferencesProvider.overrideWithValue(sharedPreferences),
      ],
      child: MyApp(),
    ),
  );
}
```


For case 1, I don't think that is an issue. After all, the value is genuinely nullable.

> in this case, I could choose to initialise the value with an `AsyncLoading` instance

`AsyncLoading` is the recommended solution

**Reply by scopendo** [reply]
Thank you – will follow those recommendations.

**Reply by Maus3rSR** [reply]
> For case 2, have the default implementation throw instead:
> 
> ```dart
> final sharedPreferencesProvider = Provider<SharedPreferences>((_) => throw UnimplementedError());
> 
> void main() async {
>   WidgetsFlutterBinding.ensureInitialized();
>   final sharedPreferences = await SharedPreferences.getInstance();
>   runApp(
>     ProviderScope(
>       overrides: [
>         sharedPreferencesProvider.overrideWithValue(sharedPreferences),
>       ],
>       child: MyApp(),
>     ),
>   );
> }
> ```
> 
> For case 1, I don't think that is an issue. After all, the value is genuinely nullable.
> 
> > in this case, I could choose to initialise the value with an `AsyncLoading` instance
> 
> `AsyncLoading` is the recommended solution

Hello,

I am curious about how would you initialize the value with an  `AsyncLoading` instance.

I am trying to create a family provider 
```dart
/// A provider that individually fetches a [WorkDetail] based on its ID.
///
/// This rarely perform an HTTP request as most of the time the [WorkDetail]
/// is already pre-fetched by the catalog.
///
/// The catch is: When using deep-links or web, a user may want to see a [WorkDetail]
/// without clicking on its item in the home page – in which case the [WorkDetail]
/// wasn't obtained yet.
/// In that situation, the provider will trigger an HTTP request to read that
/// [WorkDetail] specifically.
@riverpod
class Detail extends _$Detail {
  /// The base state is null
  @override
  FutureOr<WorkDetail?> build(String id) async {
    state = const AsyncValue.loading();
    return null;
  }

  /// Generally used by other providers which had already
  /// prefetched the same datas and want to pass them
  /// before arriving into the DetailScreen
  replace(WorkDetail detail) {
    state = AsyncValue.data(detail);
    ref.keepAlive();
  }
}
```

Because I would like to pass prefetch data from another provider (I do not cache data from repository as you did in Marvel example) or executing the fetching from my screen widget.

Take you for your time. Riverpod is really awesome...

**Reply by Maus3rSR** [reply]
By the way, I understood taht build will erase the state and that's why AsyncValue.loading will be erased. That's why I'm asking how to achieve this.

Or maybe I can override a classic provider with the prefetch value but I do not know how to achieve this override from another provider. I'm thinking it's not really possible?
